### PR TITLE
[SPARK-40280][SQL] Add support for parquet push down for annotated int and long

### PR DESCRIPTION
**Pull request by revans2**
### What changes were proposed in this pull request?
This fixes SPARK-40280 by normalizing a parquet int/long that has optional metadata with it to look like the expected version that does not have the extra metadata.

## Why are the changes needed?
This allows predicate push down in parquet to work when reading files that are complaint with the parquet specification, but different from what Spark writes.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
I added unit tests that cover this use case. I also did some manual testing on some queries to verify that less data is actually read after this change.

### CODE REVIEW

**Review by tgravescs — APPROVED** [review]
lgtm. This make sense to me based on the wording in the spec.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala [hunk 1]
```diff
@@ -76,7 +76,6 @@ import org.apache.spark.util.{AccumulatorContext, AccumulatorV2, Utils}
  * within the test.
  */
 abstract class ParquetFilterSuite extends QueryTest with ParquetTest with SharedSparkSession {
-
```
Comment (huaxingao) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala:79]: nit: unnecessary change?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala [hunk 2]
```diff
@@ -370,6 +369,38 @@ abstract class ParquetFilterSuite extends QueryTest with ParquetTest with Shared
     }
   }
 
+  test("filter pushdown - int SPARK-40280") {
```
Comment (huaxingao) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala:372]: nit: maybe `test("SPARK-40280: filter pushdown - int with annotation")`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala [hunk 3]
```diff
@@ -370,6 +369,38 @@ abstract class ParquetFilterSuite extends QueryTest with ParquetTest with Shared
...
+    checkFilterPredicate(Literal(1) === intAttr, classOf[Eq[_]], 1)
+    checkFilterPredicate(Literal(1) <=> intAttr, classOf[Eq[_]], 1)
+    checkFilterPredicate(Literal(2) > intAttr, classOf[Lt[_]], 1)
+    checkFilterPredicate(Literal(3) < intAttr, classOf[Gt[_]], 4)
+    checkFilterPredicate(Literal(1) >= intAttr, classOf[LtEq[_]], 1)
+    checkFilterPredicate(Literal(4) <= intAttr, classOf[GtEq[_]], 4)
+
+    checkFilterPredicate(!(intAttr < 4), classOf[GtEq[_]], 4)
+    checkFilterPredicate(intAttr < 2 || intAttr > 3, classOf[Operators.Or],
+      Seq(Row(1), Row(4)))
```
Comment (huaxingao) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala:401]: nit: shall we also test `In` ?

**Comment by huaxingao** [comment]
The changes look good to me.

**Comment by HyukjinKwon** [comment]
cc @wangyum FYI

**Comment by HyukjinKwon** [comment]
looks making sense to me too

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 4]
```diff
@@ -59,6 +59,21 @@ class ParquetFilters(
   // nested columns. If any part of the names contains `dots`, it is quoted to avoid confusion.
   // See `org.apache.spark.sql.connector.catalog.quote` for implementation details.
   private val nameToParquetField : Map[String, ParquetPrimitiveField] = {
+    def getNormalizedLogicalType(p: PrimitiveType): LogicalTypeAnnotation = {
+      // signed 64 bits on an INT64 and signed 32 bits on an INT32 are optional, but the rest of
+      // the code here assumes they are not set, so normalize them to not being set. SPARK-40280
+      p.getLogicalTypeAnnotation match {
+        case la : IntLogicalTypeAnnotation if la.isSigned &&
```
Comment (sadikovi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala:66]: nit: `case la: IntLogicalTypeAnnotation`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 5]
```diff
@@ -59,6 +59,21 @@ class ParquetFilters(
...
+            la.getBitWidth == 32 && p.getPrimitiveTypeName == PrimitiveTypeName.INT32 =>
+          null
+        case la : IntLogicalTypeAnnotation if la.isSigned &&
```
Comment (sadikovi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala:69]: nit: `case la: IntLogicalTypeAnnotation`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala (hunk 4 above)
Comment (sadikovi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala:66]: Also, what does "la" stand for?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 6]
```diff
@@ -59,6 +59,21 @@ class ParquetFilters(
   // nested columns. If any part of the names contains `dots`, it is quoted to avoid confusion.
   // See `org.apache.spark.sql.connector.catalog.quote` for implementation details.
   private val nameToParquetField : Map[String, ParquetPrimitiveField] = {
+    def getNormalizedLogicalType(p: PrimitiveType): LogicalTypeAnnotation = {
+      // signed 64 bits on an INT64 and signed 32 bits on an INT32 are optional, but the rest of
```
Comment (sadikovi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala:63]: nit: `SPARK-40280: Signed 64 ...`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 7]
```diff
@@ -59,6 +59,21 @@ class ParquetFilters(
...
+      // the code here assumes they are not set, so normalize them to not being set. SPARK-40280
```
Comment (sadikovi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala:64]: nit: Remove `SPARK-40280` at the end.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 8]
```diff
@@ -59,6 +59,21 @@ class ParquetFilters(
...
+      p.getLogicalTypeAnnotation match {
+        case la : IntLogicalTypeAnnotation if la.isSigned &&
+            la.getBitWidth == 32 && p.getPrimitiveTypeName == PrimitiveTypeName.INT32 =>
+          null
+        case la : IntLogicalTypeAnnotation if la.isSigned &&
+            la.getBitWidth == 64 && p.getPrimitiveTypeName == PrimitiveTypeName.INT64 =>
```
Comment (sadikovi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala:70]: I would suggest rewriting it match statement like this, IMHO cleaner: 
```scala
(p.getPrimitiveTypeName, p.getLogicalTypeAnnotation) match {
  case (INT32, intType: IntLogicalTypeAnnotation)
    if intType.getBitWidth() == 32 && intType.isSigned() => null
  case (INT64, intType: IntLogicalTypeAnnotation)
    if intType.getBitWidth() == 64 && intType.isSigned() => null
  case (_, otherType) => otherType
}
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala (hunk 2 above)
Comment (sadikovi) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala:372]: +1

**Review by sadikovi — CHANGES_REQUESTED** [review]
Looks good. I left a few comments, would appreciate it if you could address them.

**Comment by revans2** [comment]
@huaxingao and @sadikovi I think I have addressed all of your review comments. Please take another look.

**Review by huaxingao — APPROVED** [review]

**Comment by huaxingao** [comment]
LGTM. Thanks a lot for working on this! @revans2

**Review by sadikovi — APPROVED** [review]
Thanks for making the changes!

**Review by wangyum — APPROVED** [review]

**Comment by tgravescs** [comment]
merged to master, branch-3.3 and branch-3.2.  Thanks @revans2

**Pull request closed without merging**